### PR TITLE
WFCORE-3501 Allow System properties to be set at parse time before ex…

### PR DESCRIPTION
**Pull request by stuartwdouglas**
…tensions are loaded in standalone mode

### CODE REVIEW

File: controller/src/main/java/org/jboss/as/controller/parsing/DeferredExtensionContext.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
+/*
+ * JBoss, Home of Professional Open Source
+ * Copyright 2018 Red Hat Inc. and/or its affiliates and other contributors
+ * as indicated by the @authors tag. All rights reserved.
+ * See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ *
+ * This copyrighted material is made available to anyone wishing to use,
+ * modify, copy, or redistribute it subject to the terms and conditions
+ * of the GNU Lesser General Public License, v. 2.1.
```
Comment (ctomc) [inline controller/src/main/java/org/jboss/as/controller/parsing/DeferredExtensionContext.java]: this should be APL not LGPL.

**Comment by dmlloyd** [comment]
Maybe it would be simpler if _all_ properties were set at parse time, _except_ for vault expressions which are set later in the second pass?

File: server/src/main/java/org/jboss/as/server/controller/resources/SystemPropertyResourceDefinition.java [hunk 2]
```diff
@@ -69,27 +70,35 @@
...
 
+    public static final SimpleAttributeDefinition PARSE_TIME = SimpleAttributeDefinitionBuilder.create(ModelDescriptionConstants.PARSE_TIME, ModelType.BOOLEAN, true)
+            .setValidator(new ModelTypeValidator(ModelType.BOOLEAN, true))
+            .setDefaultValue(new ModelNode(false))
+            .setAllowExpression(false)
+            .build();
+
     static final AttributeDefinition[] ALL_ATTRIBUTES = new AttributeDefinition[] {VALUE, BOOT_TIME};
-    static final AttributeDefinition[] SERVER_ATTRIBUTES = new AttributeDefinition[] {VALUE};
+    static final AttributeDefinition[] SERVER_ATTRIBUTES = new AttributeDefinition[] {VALUE, PARSE_TIME};
```
Comment (bstansberry) [inline server/src/main/java/org/jboss/as/server/controller/resources/SystemPropertyResourceDefinition.java]: A reason to drop this attribute and always try to set as David suggested is it's really only valid for standalone servers, not domain servers. For domain it's cruft.

**Comment by bstansberry** [comment]
Can ExtensionXml allow a null DeferredExtensionContext? If it's null the parsing method creates a local one and then calls load() on it before returning.

That saves all the passing around of an irrelevant object for the DomainXml, HostXml and older StandaloneXml variants, plus we don't have to coordinate getting appclient in full WF changed at the same time we merge this.

**Comment by stuartwdouglas** [comment]
Retest this please

File: server/src/main/resources/schema/wildfly-config_6_0.xsd [hunk 3]
```diff
@@ -130,7 +130,7 @@
         <xs:complexType>
             <xs:sequence>
                 <xs:element name="extensions" type="extensionsType" minOccurs="0" maxOccurs="1"/>
-                <xs:element name="system-properties" type="properties" minOccurs="0" maxOccurs="1"/>
+                <xs:element name="system-properties" type="properties-with-parse-time" minOccurs="0" maxOccurs="1"/>
```
Comment (bstansberry) [inline server/src/main/resources/schema/wildfly-config_6_0.xsd]: The changes to this file should be dropped.

**Pull request merged**